### PR TITLE
Add blockchain:analyze/1 and blockchain:repair/1

### PR DESCRIPTION
**Pull request by Vagabond**
There are many things that go wrong when syncing is interrupted by some
kind of hard power down (OTAs or unplugging). This patch intends to try
to add some repair facilities for the common scenarios observed during
sync corruption.

The common sync failures are:

* Missing blocks
* Ledger inconsistency

To detect these failures several checks have been added:

* Check the delayed ledger is the right number of blocks behind
* Check the blockchain and the ledger agree on the blockchain height
* Check the delayed ledger advanced to the leading ledger fingerprints
  the same as the leading ledger

And several repair tools have been added:

* Advance the lagging ledger
* Re-sync a missing block (only gap one block wide is doable right now)
* Apply all the differences from an advanced lagging ledger to the
  leading ledger

To analyze the problems, use blockchain:analyze. Analyze will report a
single problem at a time.

To repair the current analyzed problem (if it is one we can repair), use
blockchain:repair. It will only repair a single problem at a time. It is
recommended you use analyze after each repair to check if there are more
problems, and to repair them if needed.

### CODE REVIEW

**Comment by Vagabond** [comment]
This PR also fixes a silly bug when we did not report we were doing an 'assume valid' sync and the sync height looked like it was going backwards.

File: src/blockchain.erl [hunk 1]
```diff
@@ -465,14 +493,40 @@ add_block(Block, Blockchain) ->
...
+add_block(Block, #blockchain{db=DB, blocks=BlocksCF, heights=HeightsCF, default=DefaultCF} = Blockchain, Syncing) ->
     blockchain_lock:acquire(),
     try
-        case persistent_term:get(?ASSUMED_VALID, undefined) of
-            undefined ->
-                add_block_(Block, Blockchain, Syncing);
-            AssumedValidHashAndHeight ->
-                add_assumed_valid_block(AssumedValidHashAndHeight, Block, Blockchain, Syncing)
+        PrevHash = blockchain_block:prev_hash(Block),
+        case rocksdb:get(DB, DefaultCF, ?MISSING_BLOCK, []) of
```
Comment (evanmcc) [inline src/blockchain.erl]: could we consider walling this off behind a config var?

File: src/blockchain.erl [hunk 2]
```diff
@@ -233,17 +236,23 @@ head_hash(#blockchain{db=DB, default=DefaultCF}) ->
 
 -spec sync_hash(blockchain()) -> {ok, blockchain_block:hash()} | {error, any()}.
 sync_hash(Blockchain=#blockchain{db=DB, default=DefaultCF}) ->
-    case persistent_term:get(?ASSUMED_VALID, undefined) of
-        undefined ->
-            ?MODULE:head_hash(Blockchain);
+    case rocksdb:get(DB, DefaultCF, ?MISSING_BLOCK, []) of
```
Comment (evanmcc) [inline src/blockchain.erl]: when we sync this and we get repaired, what happens with block + 1?

File: src/blockchain.erl [hunk 3]
```diff
@@ -904,6 +958,218 @@ last_block_add_time(#blockchain{default=DefaultCF, db=DB}) ->
...
+                                                                                ({deleted, _, _}, Acc) ->
+                                                                                     maps:update_with(deleted, fun(V) -> V+1 end, 1, Acc)
+                                                                             end, #{}, Changes),
+                                                           {M, maps:to_list(Sum)}
+                                                   end, Mismatches)}};
+      E ->
+          E
+  end.
+
+%% TODO it'd be nice if the fingerprint function was consistent with the column family names
```
Comment (evanmcc) [inline src/blockchain.erl]: yeah.

File: src/transactions/blockchain_txn.erl [hunk 4]
```diff
@@ -521,26 +526,40 @@ absorb_delayed(Block0, Chain0) ->
...
+                                              absorb_delayed_(Block1, Chain1);
+                                         (_, Acc) ->
+                                              Acc
+                                      end,
+                                      ok,
+                                      lists:seq(DelayedHeight+1, DelayedHeight + Lag)),
+                    case Res of
+                        ok ->
+                            ok = blockchain_ledger_v1:commit_context(DelayedLedger1);
+                        Error ->
```
Comment (evanmcc) [inline src/transactions/blockchain_txn.erl]: do we need to clear the context here?

Comment (Vagabond) [reply src/transactions/blockchain_txn.erl]: No, the error path discards the context

File: src/blockchain.erl [hunk 5]
```diff
@@ -372,12 +389,23 @@ ledger_at(Height, Chain0) ->
 
 fold_chain(Chain0, DelayedHeight, DelayedLedger, Height) ->
     lists:foldl(
-      fun(H, ChainAcc) ->
-              {ok, Block} = ?MODULE:get_block(H, Chain0),
-              {ok, Chain1} = blockchain_txn:absorb_block(Block, ChainAcc),
-              Chain1
+      fun(H, {ok, ChainAcc}) ->
```
[the remark this answers is not in the document]
Comment (evanmcc) [reply src/blockchain.erl]: ignore this

File: src/blockchain.erl (hunk 2 above)
Comment (Vagabond) [reply src/blockchain.erl]: It will go back to the previous HEAD now that the hole is filled in

File: src/blockchain.erl (hunk 1 above)
Comment (Vagabond) [reply src/blockchain.erl]: What kind of config var, application env, command line flag?

File: src/blockchain.erl (hunk 5 above)
[the remark this answers is not in the document]
Comment (Vagabond) [reply src/blockchain.erl]: Ironically I thought throwing would be more controversial

File: src/blockchain.erl (hunk 2 above)
Comment (evanmcc) [reply src/blockchain.erl]: the sync will just error out when the block after the missing one is reapplied?

File: src/blockchain.erl (hunk 1 above)
Comment (evanmcc) [reply src/blockchain.erl]: application env, so we can edit the config and restart

File: src/blockchain.erl (hunk 5 above)
[the remark this answers is not in the document]
Comment (evanmcc) [reply src/blockchain.erl]: it's a nit, really.  throwing I think is better in places where you'd really prefer to return early.

File: src/blockchain.erl (hunk 2 above)
Comment (Vagabond) [reply src/blockchain.erl]: yeah, that is the intention

**Pull request merged**